src/middlewared/middlewared/plugins/zfs.py

```diff
@@ -586,6 +588,116 @@ def query_for_quota_alert(self):
+        This is quite a bit faster than using py-libzfs.
+        currently supports the following options:
+
+        `additional_information` - expand the amount of data
+        returned from the call.
+
+        `SIZE` - size of zvol
+        `DEVID` - the device id of the zvol
+        `RO` - whether zvol is flagged as ro (snapshot)
+        `ATTACHMENT` - where zvol is currently being used
+
+        sample option:
+        {"additional_information": ["DEVID"]}
+        """
+        def get_size(zvol_dev):
+            with open(f'/sys/block/{zvol_dev}/size', 'r') as f:
+                nblocks = f.readline()
+
+            return int(nblocks[:-1]) * 512
+
+        def get_devid(zvol_dev):
+            with open(f'/sys/block/{zvol_dev}/dev', 'r') as f:
+                devid = f.readline()
+            return devid[:-1]
+
+        def get_ro(zvol_dev):
+            with open(f'/sys/block/{zvol_dev}/ro', 'r') as f:
+                ro = f.readline()
+            return ro[:-1] == '1'
+
+        def get_attachment(zvol_vdev, data):
+            out = None
+            for method, attachment in data.items():
+                val = attachment.pop(zvol_vdev, None)
+                if val is not None:
+                    out = {
+                        "method": method,
+                        "data": val
+                    }
+                    break
+
+            return out
+
+        def get_zvols(info_level):
+            out = {}
+            zvol_path = '/dev/zvol/'
+            offset = len(zvol_path)
+            iscsi_offset = len('zvol/')
+            do_get_size = "SIZE" in info_level
+            do_get_dev = "DEVID" in info_level
+            do_get_ro = "RO" in info_level
+            do_get_attachment = "ATTACHMENT" in info_level
+
+            if do_get_attachment:
+                """
+                Normalize all places where we store zvol info so that
+                indexed by dataset name for quick lookup
+                """
+                extents = self.middleware.call_sync('iscsi.extent.query', [('type', '=', 'DISK')])
+                iscsi_zvols = {
+                    i['path'][iscsi_offset:]: i for i in extents
+                }
+
+                vm_devices = self.middleware.call_sync('vm.device.query', [["dtype", "=", "disk"]])
+                vm_zvols = {
+                    zvol_path_to_name(i['attributes']['path']): i for i in vm_devices
+                }
+                attachment_data = {
+                    "iscsi.extent.query": iscsi_zvols,
+                    "vm.devices.query": vm_zvols
+                }
+
+            for root, dirs, files in os.walk(zvol_path):
+                if not files:
+                    continue
+
+                for file in files:
+                    path = root + "/" + file
+                    zvol_name = path[offset:]
+                    dev_name = os.readlink(path).split('/')[-1]
+
+                    out.update({
+                        zvol_name: {
```

Comment:
Btw just to be sure, how are we handling the spaces in the zvol name here ? Does it show them or does it use the escaped version ?

Reply:
Improved output to clearly differentiate between `name` and `path`:

```
root@truenas[~]# midclt call zfs.dataset.unlocked_zvols_fast | jq
{
  "tank space/zvol with lots of spaces and trailing space ": {
    "path": "/dev/zvol/tank+space/zvol+with+lots+of+spaces+and+trailing+space+",
    "name": "tank space/zvol with lots of spaces and trailing space ",
    "dev": "zd48"
  },
```

VM disk choices:
```
root@truenas[~]# midclt call vm.device.disk_choices | jq
{
  "/dev/zvol/tank+space/zvol+with+lots+of+spaces+and+trailing+space+": "tank space/zvol with lots of spaces and trailing space ",
```

ISCSI disk choices:
```
root@truenas[~]# midclt call iscsi.extent.disk_choices | jq
{
  "zvol/tank+space/zvol+with+lots+of+spaces+and+trailing+space+": "tank space/zvol with lots of spaces and trailing space  (1.02 MiB)",
  ```

Reply:
Confirmed out-of-band that above is correct.